utilities/transactions/transaction_base.cc

```diff
@@ -192,7 +192,53 @@ Status TransactionBaseImpl::PopSavePoint() {
   }
 
   assert(!save_points_->empty());
-  save_points_->pop();
+  // If there is another savepoint A below the current savepoint B, then A needs
+  // to inherit tracked_keys in B so that if we rollback to savepoint A, we
+  // remember to unlock keys in B. If there is no other savepoint below, then we
+  // can safely discard savepoint info.
+  if (save_points_->size() > 1) {
+    TransactionBaseImpl::SavePoint top;
+    std::swap(top, save_points_->top());
+    save_points_->pop();
+
+    const TransactionKeyMap& curr_key_map = top.new_keys_;
+    TransactionKeyMap& prev_key_map = save_points_->top().new_keys_;
+
+    for (const auto& key_map_iter : curr_key_map) {
+      uint32_t column_family_id = key_map_iter.first;
+      auto& keys = key_map_iter.second;
+
+      // If cfid was not previously tracked, just copy everything over.
+      auto prev_tracked_keys = prev_key_map.find(column_family_id);
```

Comment:
I think it would make it clear if auto is replaced with the explicit type here.

Reply:
Since this is a std::pair, I think I'll leave it as auto since the type is quite long. Hopefully the explicit types elsewhere will help the reader.